article/fivetran custom connector (#226)
MechanicalRock/mechanicalrock.github.io: _posts/2022-04-26-fivetran-custom-connector.md

---
layout: post
title: FiveTran Paginated Custom Connector
date: 2022-05-26

> tags: fivetran custom connector custom-connector pagination getmore state management

Author: getmore is a tag people will search for.

agree on state-management

> tags: fivetran custom connector custom-connector pagination getmore state management

Author: ```suggestion
tags: fivetran custom connector custom-connector pagination getmore state-management
```

author: Joseph Miocevich
image: img/fivetran_logo.png
description:
  Overview of a custom fivetran connector with pagination
---

![Header](/img/fivetran_logo.png)

> <br/>

Author: Added

## Overview

> In simple terms, a FiveTran custom connector a cloud hosted function that can be setup on a cloud platform of your choice, which allows you to load data into FiveTran with a API FiveTran does not natively support. The benefits of this is that these connectors will have the same capabilities of standard connectors:

Reviewer: (if this is a quote, disregard)

> In simple terms, a FiveTran custom connector a cloud hosted function that can be setup on a cloud platform of your choice, which allows you to load data into FiveTran with a API FiveTran does not natively support. The benefits of this is that these connectors will have the same capabilities of standard connectors:

Author: Its not a direct quote, but very similar, can reword

- Incremental updates
- Source data type inference
- Automatic schema updates
- Data de-deduplication
- Destination ingestion optimisations
- Logs and alerts to monitor events and troubleshoot issues

> Use FiveTran's Function connectors if:

Reviewer 2: Fivetran's Lambda connectors ?

Author: Can be any cloud connector

- Fivetran doesn't have a connector for your source
- You are using private APIs or custom applications
- You are using a source or API that Fivetran is unlikely to support in the near future
- You want to sync unsupported file formats that require pre-processing
- You have sensitive data that needs filtering or anonymizing before entering the destination

> *Fivetran* . Fivetran Custom Connector. (n.d.). Retrieved May 26, 2022, from https://fivetran.com/docs/functions 

Reviewer: Is the above section a direct quote?

If so, [make it a blockquote](https://commonmark.org/help/tutorial/05-blockquotes.html).

Author: yes, will block quote

<br>

***
<br>
## General Principles

**State Management**

State is a JSON object that contains cursors from the previous successful FiveTran executions run.

[https://fivetran.com/docs/functions/faq/use-state-object](https://fivetran.com/docs/functions/faq/use-state-object) 

[https://fivetran.com/docs/functions/faq/use-secrets-object](https://fivetran.com/docs/functions/faq/use-secrets-object)

> **Guardrails to consider**

Reviewer: Guardrails?

These read more like _potential issues_.  or just _things_.

Also - Title Case

Author: ```suggestion
**Potential Issues to Consider**
```

- Lambda timeouts and payload limit

> - Ensure calls to the Lambda are **idempotent** for a given FiveTran cursor

Reviewer: ```suggestion
- Ensure calls to the Lambda are [**idempotent**](https://en.wikipedia.org/wiki/Idempotence) for a given FiveTran cursor
```

Reader may not be familiar with the term

Author: Good point

- Be wary or avoid using cursors supplied by the pagination of an upstream API that may not be persistent.

**FiveTran Pagination**

FiveTran pagination allows your lambda function to specify if there is more data to be collected, this is achieved using the hasMore boolean return.

> When hasMore = True ,state is updated as normal, however FiveTran immediately calls the lambda with the updated state. This will keep occurring until hasMore = False, which then resets FiveTran to its default state.

Reviewer 2: Worth mentioning if it's in a default state, the next sync will be based on the sync frequency set

Author: Good point, will add in

> When hasMore = True ,state is updated as normal, however FiveTran immediately calls the lambda with the updated state. This will keep occurring until hasMore = False, which then resets FiveTran to its default state.

Reviewer: ```suggestion
When `hasMore = True` ,state is updated as normal, however FiveTran immediately calls the lambda with the updated state. This will keep occurring until `hasMore = False`, which then resets FiveTran to its default state.
```

[https://fivetran.com/docs/functions/faq/use-hasmore-flag](https://fivetran.com/docs/functions/faq/use-hasmore-flag)

Example Expected function payload

```
{
      state: {
        cursor: '2020-01-01',
        paginationCounter: 0
      },
      secrets: Object
    }
```

Example Expected Connector response format

```
state: {
cursor: cursorPoint,
paginationCounter: 1
    },
    insert: {
      Table: apiResponseJson
    },
    schema: {
      Table: {
        primary_key: ['some_unique_key']
      }
    },
    hasMore: boolean
  }
}

```

<br/>

***

<br/>
## Sequence of FiveTran Connector Excecutions

![](/img/five_tran_sequence.png)

For the sake of this example and readability, the return JSON from the api call will be summarised as *apiJsonResponse*

### Call 1: Initial Sync

**FiveTranCall**

```
state: {}
```

**Connector Response**

> ```

Reviewer: format the json for readability,

you might also be able [to style it](https://stackoverflow.com/questions/14901245/how-to-style-a-json-block-in-github-wiki) (YMMV)

Author: I originally had it as json but didn't like the way it looked, but that link gives more options, will implement

Author: updated

{
state: {
lastUpdate: ''
paginationCounter: 1
},
insert: {
apiJsonResponseTable: apiJsonResponse
},
schema: {
apiJsonResponseTable: {
primary_key: ['id']
}
},
hasMore: True
}
```

> Initial API sync with no state, api call is getting all data with no state and has returned a paginated response

Reviewer: a bit clunky - [read it out loud](https://docs.google.com/document/d/1W1e0nKku0IYZ5spUXIeI8TG04sAELGvT67CHGHGwJT0/edit#heading=h.2grqrue)

Author: Agreed

```suggestion
The initial FiveTran sync contains no state, triggering our custom connector to fetch all data from the data source API. The API has returned page 1/3 of a paginated response
```

> Key response features

Reviewer: This section is good - it's worth signposting it better - either with a heading, or bold text.

Author: ```suggestion
**Key response features**
```

State: 
- No LastUpdate as paginated query has not been completed
- Pagination Counter incremented + 1

> - hasMore = True

Reviewer: ```suggestion
- `hasMore = True`
```

<br/>
### Call 2: Paginated Query 1

**FiveTranCall**

> ```

Reviewer: format json

state: {
lastUpdate: '',
paginationCounter: 1
}
```

> **Connector Response**

Reviewer 2: Custom Connector Response / Lambda Connector Response ?

Author: I considered having both, I originally had custom connector response but thought it was unnecessary given the context

This isn't AWS specific so don't want to put lambda

> **Connector Response**

Reviewer: Is this intended to be a heading?

> ```

Reviewer: fmt

{
state: {
lastUpdate: '',
paginationCounter: 2
},
insert: {
apiJsonResponseTable: apiJsonResponse
},
schema: {
apiJsonResponseTable: {
primary_key: ['id']
}
},
hasMore: True
}
```

Paginating through the api response for initial sync

> Key response features

Reviewer: see above

Author: ```suggestion
**Key response features**
```

State: 
- No LastUpdate as paginated query has not been completed
- Pagination Counter incremented + 1
- hasMore = True

<br/>
### Call 3: Paginated Query 2

**FiveTranCall**

```

> state: {

Reviewer: ALT + SHIFT + F

lastUpdate: '',
paginationCounter: 2
}
```

**Connector Response**

> ```

Reviewer: :point_up:

- Pagination Counter set to 0, as paginated query is completed
- hasMore = false

<br/>
### Call 4: Next FiveTran Sync

**FiveTranCall**

```
state: {
lastUpdate: '2020-01-01',
paginationCounter: 0
}
```

**Connector Response**

```
{
state: {
lastUpdate: '2020-01-02',
paginationCounter: 0
},
insert: {
apiJsonResponseTable: apiJsonResponse
},
schema: {
apiJsonResponseTable: {
primary_key: ['id']
}
},
hasMore: False
}
```

FiveTran Sync with LastUpdate State

Key response features
State: 
- LastUpdate updated, query has been completed
- Pagination Counter set to 0, as no pagination necessary 
- hasMore = false

<br/>
### Call 5: Next FiveTran Sync

**FiveTranCall**

```
state: {
lastUpdate: '2020-01-02',
paginationCounter: 0
}
```

**Connector Response**

```

state: {
lastUpdate: '2020-01-03',
paginationCounter: 0
},
insert: {
apiJsonResponseTable: apiJsonResponse
},
schema: {
apiJsonResponseTable: {
primary_key: ['id']
}
},
hasMore: False
})
}
```

FiveTran Sync with LastUpdate State

Key response features
State: 

- LastUpdate updated, query has been completed
- Pagination Counter set to 0, as no pagination necessary 
- hasMore = false

<br/>

***

<br/>

### **Examples**

> For full working example see FiveTran's documentation

Reviewer: ```suggestion
For a full working example see FiveTran's documentation
```

---
layout: post
title: FiveTran Paginated Custom Connector
date: 2022-05-26
tags: fivetran custom connector custom-connector pagination getmore state-management
author: Joseph Miocevich
image: img/fivetran_logo.png
description:
  Overview of a custom fivetran connector with pagination
---

![FiveTran Paginated Custom Connectors](/img/fivetran_logo.png)

<br/>

> Understanding and debugging your FiveTran custom connector is not straightfoward, with many online tutorials excluding the key component of data API's... pagination. After reading this blog you will have a better understanding of the expected behaviour between FiveTran and your custom connector when pagination is thrown into the mix.

Reviewer: :+1:

## Overview

A FiveTran custom connector is a cloud hosted function that can be set up on a cloud platform of your choice, which allows you to load data into FiveTran with a API FiveTran does not natively support. 

Like standard connectors, custom connectors have a number of benefits:

> >

Reviewer: Does this format the quote properly?  Thought it needed to be on each line.

Author: Formats properly :)

- Incremental updates
- Source data type inference
- Automatic schema updates
- Data de-deduplication
- Destination ingestion optimisations
- Logs and alerts to monitor events and troubleshoot issues

Use FiveTran's Function connectors if:
>
- Fivetran doesn't have a connector for your source
- You are using private APIs or custom applications
- You are using a source or API that Fivetran is unlikely to support in the near future
- You want to sync unsupported file formats that require pre-processing
- You have sensitive data that needs filtering or anonymizing before entering the destination

*Fivetran* . Fivetran Custom Connector. (n.d.). Retrieved May 26, 2022, from https://fivetran.com/docs/functions 

<br>

***
<br>
## General Principles

**State Management**

State is a JSON object that contains cursors from the previous successful FiveTran executions run.

[https://fivetran.com/docs/functions/faq/use-state-object](https://fivetran.com/docs/functions/faq/use-state-object) 

[https://fivetran.com/docs/functions/faq/use-secrets-object](https://fivetran.com/docs/functions/faq/use-secrets-object)

**Potential Issues to Consider**

> - Lambda timeouts and payload limit

Reviewer: Add link to the relevant AWS documentation?

Author: Good idea

Author: added

Example Expected function payload

```javascript
{
      state: {
        cursor: '2020-01-01',
        paginationCounter: 0
      },
      secrets: Object
    }
```

Example Expected Connector response format

```javascript
state: {
cursor: cursorPoint,
paginationCounter: 1
    },
    insert: {
      Table: apiResponseJson
    },
    schema: {
      Table: {
        primary_key: ['some_unique_key']
      }
    },
    hasMore: boolean
  }
}

```

<br/>

***

<br/>
## Sequence of FiveTran Connector Excecutions

![](/img/five_tran_sequence.png)

For the sake of this example and readability, the return JSON from the api call will be summarised as *apiJsonResponse*

### Call 1: Initial Sync

**FiveTranCall**

```javascript
state: {}
```

**Connector Response**

```javascript
{
state: {
lastUpdate: ''
paginationCounter: 1
},
insert: {
apiJsonResponseTable: apiJsonResponse
},
schema: {
apiJsonResponseTable: {
primary_key: ['id']
}
},
hasMore: True
}
```

Initial API sync with no state, API call is getting all data with no state and has returned a paginated response

**Key response features**
State: 

> - No LastUpdate as paginated query has not been completed

Reviewer: ```suggestion
- No `lastUpdate` as paginated query has not been completed
```

- Pagination Counter incremented + 1
- `hasMore = True`

<br/>
### Call 2: Paginated Query 1

**FiveTranCall**

```javascript
state: {
lastUpdate: '',
paginationCounter: 1
}
```

**Connector Response**

```javascript
{
state: {
lastUpdate: '',
paginationCounter: 2
},
insert: {
apiJsonResponseTable: apiJsonResponse
},
schema: {
apiJsonResponseTable: {
primary_key: ['id']
}
},
hasMore: True
}
```

Paginating through the API response for initial sync

**Key response features**
State: 
- No LastUpdate as paginated query has not been completed
- Pagination Counter incremented + 1
- hasMore = True

<br/>
### Call 3: Paginated Query 2

**FiveTranCall**

```javascript
state: {
lastUpdate: '',
paginationCounter: 2
}
```

**Connector Response**

```javascript
{
state: {
lastUpdate: '2020-01-01',
paginationCounter: 0
},
insert: {
apiJsonResponseTable: apiJsonResponse
},
schema: {
apiJsonResponseTable: {
primary_key: ['id']
}
},
hasMore: False
}
```

Paginating through the API response for initial sync

> Key response features

Reviewer: ```suggestion
**Key response features**
```

consistent formatting

State: 

> - LastUpdate Set as paginated query has been completed

Reviewer: ```suggestion
- `lastUpdate` Set as paginated query has been completed
```

maintain consistency with code

- Pagination Counter set to 0, as paginated query is completed
- hasMore = false

<br/>
### Call 4: Next FiveTran Sync

**FiveTranCall**

```javascript
state: {
lastUpdate: '2020-01-01',
paginationCounter: 0
}
```

**Connector Response**

```javascript
{
state: {
lastUpdate: '2020-01-02',
paginationCounter: 0
},
insert: {
apiJsonResponseTable: apiJsonResponse
},
schema: {
apiJsonResponseTable: {
primary_key: ['id']
}
},
hasMore: False
}
```

FiveTran Sync with LastUpdate State

> Key response features

Reviewer: ```suggestion
**Key response features**
```

State: 
- LastUpdate updated, query has been completed
- Pagination Counter set to 0, as no pagination necessary 
- hasMore = false

<br/>
### Call 5: Next FiveTran Sync

**FiveTranCall**

```javascript
state: {
lastUpdate: '2020-01-02',
paginationCounter: 0
}
```

**Connector Response**

```javascript
state: {
lastUpdate: '2020-01-03',
paginationCounter: 0
},
insert: {
apiJsonResponseTable: apiJsonResponse
},
schema: {
apiJsonResponseTable: {
primary_key: ['id']
}
},
hasMore: False
})
}
```

FiveTran Sync with LastUpdate State

> Key response features

Reviewer: ```suggestion
**Key response features**
```

State: 

> - LastUpdate updated, query has been completed

Reviewer: ```suggestion
- `lastUpdate` updated, query has been completed
```
